src/util/alloc/fd_alloc.c

```diff
@@ -0,0 +1,942 @@
+
+ulong
+fd_alloc_align( void ) {
+  return alignof(fd_alloc_t);
+}
+
+ulong
+fd_alloc_footprint( void ) {
+  return sizeof(fd_alloc_t);
+}
+
+void *
+fd_alloc_new( void * shmem ) {
+
+  /* Check input arguments */
+
+  if( FD_UNLIKELY( !shmem ) ) {
+    FD_LOG_WARNING(( "NULL shmem" ));
+    return NULL;
+  }
+
+  if( FD_UNLIKELY( !fd_ulong_is_aligned( (ulong)shmem, alignof(fd_alloc_t) ) ) ) {
+    FD_LOG_WARNING(( "misaligned shmem" ));
+    return NULL;
+  }
+
+  fd_wksp_t * wksp = fd_wksp_containing( shmem );
+  if( FD_UNLIKELY( !wksp ) ) {
+    FD_LOG_WARNING(( "shmem must be in a workspace" ));
+    return NULL;
+  }
+
+  fd_alloc_t * alloc = (fd_alloc_t *)shmem;
+  fd_memset( alloc, 0, sizeof(fd_alloc_t) );
+
+  alloc->wksp_off = (ulong)alloc - (ulong)wksp;
+
+  FD_COMPILER_MFENCE();
+  FD_VOLATILE( alloc->magic ) = FD_ALLOC_MAGIC;
+  FD_COMPILER_MFENCE();
+
+  return shmem;
+}
+
+fd_alloc_t *
+fd_alloc_join( void * shalloc,
+               ulong  cgroup_idx ) {
+  fd_alloc_t * alloc = (fd_alloc_t *)shalloc;
+
+  if( FD_UNLIKELY( !alloc ) ) {
+    FD_LOG_WARNING(( "NULL shalloc" ));
+    return NULL;
+  }
+
+  if( FD_UNLIKELY( !fd_ulong_is_aligned( (ulong)alloc, alignof(fd_alloc_t) ) ) ) {
+    FD_LOG_WARNING(( "misaligned shalloc" ));
+    return NULL;
+  }
+
+  if( FD_UNLIKELY( alloc->magic!=FD_ALLOC_MAGIC ) ) {
+    FD_LOG_WARNING(( "bad magic" ));
+    return NULL;
+  }
+
+  if( FD_UNLIKELY( cgroup_idx>=FD_ALLOC_JOIN_CGROUP_CNT ) ) {
+    FD_LOG_WARNING(( "bad cgroup_idx" ));
+    return NULL;
+  }
+
+  return fd_alloc_private_join( alloc, cgroup_idx );
+}
+
+void *
+fd_alloc_leave( fd_alloc_t * join ) {
+
+  if( FD_UNLIKELY( !join ) ) {
+    FD_LOG_WARNING(( "NULL join" ));
+    return NULL;
+  }
+
+  return fd_alloc_private_join_alloc( join );
+}
+
+void *
+fd_alloc_delete( void * shalloc ) {
+
+  if( FD_UNLIKELY( !shalloc ) ) {
+    FD_LOG_WARNING(( "NULL shalloc" ));
+    return NULL;
+  }
+
+  if( FD_UNLIKELY( !fd_ulong_is_aligned( (ulong)shalloc, alignof(fd_alloc_t) ) ) ) {
+    FD_LOG_WARNING(( "misaligned shalloc" ));
+    return NULL;
+  }
+
+  fd_alloc_t * alloc = (fd_alloc_t *)shalloc;
+
+  if( FD_UNLIKELY( alloc->magic!=FD_ALLOC_MAGIC ) ) {
+    FD_LOG_WARNING(( "bad magic" ));
+    return NULL;
+  }
+
+  FD_COMPILER_MFENCE();
+  FD_VOLATILE( alloc->magic ) = ~FD_ALLOC_MAGIC;
+  FD_COMPILER_MFENCE();
+
+  /* Clean up as much as we can.  For each sizeclass, delete all active
+     superblocks and all inactive superblocks.  We will not be able
+     cleanup any superblocks that are fully allocated (and any
+     outstanding large allocations) as we rely on the application to
+     implicitly track them (because they nominally will eventually call
+     free on them).  So hopefully the application did a good job and
+     cleaned up all their outstanding stuff before calling delete. */
+
+  fd_wksp_t * wksp = fd_alloc_private_wksp( alloc );
+
+  for( ulong sizeclass=0UL; sizeclass<FD_ALLOC_SIZECLASS_CNT; sizeclass++ ) {
```

Comment:
Is there a reason you explicitly take apart your data structure when you could just memset it to zero or just reset the magic number and it is gone?

Reply:
The taking apart is required to get rid of superblocks still lying around even after the user (being righteous and proper) has cleaned up all their allocations.  If it was just zeroed, there would likely be huge superblocks stranded in the allocator of last resort.

Reply:
Note: could argue that you could limit this to just cleaning up huge superblock sizeclasses to tune it.  But this isn't a critical performance path as it is purely teardown.  And, doing it this way is useful was simpler and also was done to make it easy to do the types of instrumentation / integrity checking discussed above without impact performance.

For example, in a righteous and proper application, all the superblocks at this point found should be completely empty if deleted in the above order during the tear down.  Quite possible to test this and noisily log about it.

This noisy logging is not done currently because a common usage pattern to have a quick-and-dirty application call a bunch of mallocs and then just terminate without cleaning up (see rants again ... they aren't actually using malloc-free semantics).  Another common usage pattern is to terminate anomalously leaving things deliberately unclean to help with core dumps and what not.  In either case, a default policy of noisily logging such is not helpful and possibly self defeating in a bigger picture sense if it trains users to ignore logs.

Reply:
Ahh, I understand